### PR TITLE
Adding GPU functionality, enabling data loaders/tuples as input

### PR DESCRIPTION
**Pull request by jdunnmon**
- [x]  Added cuda functionality in `Classifier` and `EndModel`

- [x] Modifies `Classifier` and `EndModel` to give user option to use either `DataLoader` or `tuple` as input 

- [x] Refactors to support the above + small bugfixes (import typos in `contrib`, etc.)

- [x] Adds option for training progress bar for `EndModel`  

- [x] Updates multitask class structure to leverage the above

- [x] Update tests

### CODE REVIEW

**Comment by ajratner** [comment]
@jdunnmon Can we merge this into `master` instead?  I think we should use `dev` for major changes to the system, whereas I think we should get smaller functionality additions like this one (even if they change the interface a bit) into `master` as quickly as possible so people can start using them!

@bhancock8 thoughts on this?

@jdunnmon once you've done this & tests are passing, tag me here so I can review then?  Thanks!

**Comment by jdunnmon** [comment]
Yeah, I can prep the PR for `master`.  I'll need to change the tests, etc., but that's on the to do list anyway, just a bit more important to make sure everything's locked down before moving it into `master`.  I'll get that ready.

**Comment by ajratner** [comment]
Cool!  I also suggest bc looks like this PR is already pulling in changes from master into dev, so will probably be cleaner anyway!

File: metal/classifier.py [hunk 1]
```diff
@@ -271,10 +292,84 @@ def _set_scheduler(self, scheduler_config, optimizer):
...
+                Y = self._break_ties(Y, break_ties)
+
+            Y_p = self.predict(X, break_ties=break_ties, **kwargs)
+
+        elif type(data) is DataLoader:
+            Y_p, Y = self._batch_evaluate(data, break_ties=break_ties)
+
+        else:
+            raise ValueError(
+                "Unrecognized input data structure, use tuple or DataLoader!"
```
Comment (bhancock8) [inline metal/classifier.py]: Not sure we need an exclamation point. :)

Comment (jdunnmon) [reply metal/classifier.py]: Lol, sounds good.  Will update

**Comment by bhancock8** [comment]
As I've looked at it a little more, I see the issue: if they give us a dataloader with X and Y in score(), then we need to somehow split it before passing to predict(), since it expects only Xs. Here's what I suggest then:

--score() accepts either (X, Y) or a DataLoader. If it's (X, Y), then first thing we make it into a DataLoader. For each batch, it calls predict, stores the results, and then once we have all of Y_p and Y, we calculate metrics once.
--predict() accepts either X or a DataLoader. And again, first thing we convert into a DataLoader, then for each batch, call predict_proba() and do tie-breaking.
--So internally, we're always using DataLoaders (consistent with the train method); we just offer the tuple signature for convenience. And people can do various things with their DataLoaders if they so choose--shuffling, parallel loading, batching, etc. We don't care.

**Comment by ajratner** [comment]
@bhancock8 agreed in full!!  @jdunnmon ping me when adjustments made and I'll do a more thorough look then!

**Comment by jdunnmon** [comment]
@bhancock8 @ajratner thanks for the thoughts!  think that will work, i'll put in these changes (and update `MultitaskClassifier`) as soon as I can.

**Comment by bhancock8** [comment]
Thank you for putting it in! This will be good to have.

**Comment by jdunnmon** [comment]
@ajratner take a look at this -- tests should pass now.  I think there's an argument for moving `evaluate` from `classifier` down into `EndModel` and `LabelModel`, as the reason I have to duplicate code inside atm is because I put it in 'Classifier'.  Does it make more sense to put a stub in `Classifier` and move the actual implementation down the stack?

**Comment by ajratner** [comment]
@jdunnmon Looks like it is indeed the format checker... I'll look over the rest of the PR in the mean time.  Can you also resolve conflicts?

File: metal/classifier.py [hunk 2]
```diff
@@ -142,10 +144,27 @@ def _train(self, train_loader, loss_fn, X_dev=None, Y_dev=None):
...
+            if self.config["verbose"]:
+                print("Using GPU...")
+
+            self.cuda()
+
         # Train the model
         for epoch in range(train_config["n_epochs"]):
             epoch_loss = 0.0
-            for data in train_loader:
+            for batch, data in tqdm(
```
Comment (bhancock8) [inline metal/classifier.py]: 'batch' can be a lot of things--e.g., data, or a counter. In this case, it's a counter, right? So maybe, batch_num.

File: metal/classifier.py [hunk 3]
```diff
@@ -142,10 +144,27 @@ def _train(self, train_loader, loss_fn, X_dev=None, Y_dev=None):
...
+                enumerate(train_loader),
```
[the remark this answers is not in the document]
Comment (bhancock8) [reply metal/classifier.py]: I like the idea of us creating the DataLoader using config options (so batch size, shuffle, memory pinning, etc. can all be handled internally), but I think the issue would be that we're making this changes to DataLoaders to handle cases where the user can't fit their entire dataset in memory, and them passing in a Dataset would require that, I think. A DataLoader on the other hand could dynamically load things. I think as quickly as possible in our workflow we should convert to one type (DataLoader) and use that consistently everywhere, so maybe they pass either a tuple, or Dataset, or DataLoader, and that immediately becomes a DataLoader.

File: metal/classifier.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bhancock8) [reply metal/classifier.py]: Yeah, I proposed the same before--no need for evaluate. predict() handles all predicting, and score() just calls predict and calculates metrics. Batching is all handled by the DataLoader (which we either construct using their config options or which is passed in by them).

File: metal/end_model/end_model.py [hunk 4]
```diff
@@ -72,7 +71,9 @@ def _build(self, input_module, middle_modules, head_module):
             self.network = nn.Sequential(input_layer, *middle_layers, head)
 
         # Construct loss module
-        self.criteria = SoftCrossEntropyLoss(reduction="sum")
+        self.criteria = SoftCrossEntropyLoss(
+            reduction="sum", use_cuda=self.config["train_config"]["use_cuda"]
```
Comment (bhancock8) [inline metal/end_model/end_model.py]: Do we need a separate use_cuda setting for train and eval? Any reason it can't just be a single "use_cuda" for the whole model?

File: metal/classifier.py [hunk 5]
```diff
@@ -271,10 +292,84 @@ def _set_scheduler(self, scheduler_config, optimizer):
...
+        if type(data) is tuple:
+            X, Y = data
+
+            if self.config["train_config"]["use_cuda"]:
+                X = X.cuda()
+
+            Y = self._to_numpy(Y)
+
+            if Y.ndim > 1:
+                Y = self._break_ties(Y, break_ties)
```
[the remark this answers is not in the document]
Comment (bhancock8) [reply metal/classifier.py]: Agreed--if it's called Y and not Y_p, then it's integer values; no ties to break.

[the remark this answers is not in the document]
Comment (ajratner) [reply metal/classifier.py]: Yeah so apparently we're passing these in as one-hots in some of the tests- I'm going to change it to ints.  `Y_train` might be in vector form, but in single task, `Y` used for eval should be ints (we should also be clear about this)

[the remark this answers is not in the document]
Comment (ajratner) [reply metal/classifier.py]: Nvm just due to changes below

**Comment by ajratner** [comment]
@jdunnmon @bhancock8 Done with my pass here!  Take a look when you get a chance?

Only remaining things that would be nice to have:
- [ ] A test subdirectory for simple GPU functionality testing (not to be run on Travis, but so that a user can check themselves)
- [ ] Nicer way to handle putting our loss function on CUDA?

**Comment by ajratner** [comment]
Oh also, my tests keep failing due to `make check`, even though this passes locally (same as JD was seeing).  @bhancock8 any ideas here?

File: metal/end_model/loss.py [hunk 6]
```diff
@@ -18,17 +18,24 @@ class SoftCrossEntropyLoss(nn.Module):
...
 
-    def __init__(self, weight=None, reduction="elementwise_mean"):
+    def __init__(
+        self, weight=None, reduction="elementwise_mean", use_cuda=False
+    ):
         super().__init__()
         assert weight is None or isinstance(weight, torch.FloatTensor)
         self.weight = weight
         self.reduction = reduction
+        self.use_cuda = use_cuda
```
[the remark this answers is not in the document]
Comment (jdunnmon) [reply metal/end_model/loss.py]: @ajratner my remembrance is that it didn't work because of these two lines:

`cum_losses = torch.zeros(n)`
`cls_idx = torch.full((n,), y, dtype=torch.long)`

each of which define a new variable inside the module, which by default was on the CPU.  I manually switched that to GPU to make sure that did not cause a problem.  We also put the rest of the module on the GPU as well when we do `criteria.cuda()` in `get_loss_fn`.  Make sense?

[the remark this answers is not in the document]
Comment (ajratner) [reply metal/end_model/loss.py]: Got it!

**Comment by ajratner** [comment]
@bhancock8 @jdunnmon NVM, I'm pretty sure the formatting errors are coming from master.  So should be fine to merge.  @bhancock8 when you're up in a few hours, let us know if you can quickly review and approve the changes?

**Comment by jdunnmon** [comment]
@bhancock8 , ready for your eyes.  @ajratner and I went through today and made a bunch of changes to update the above and added a test for successful GPU usage.  Go ahead and give it a look -- want to make any fixes tomorrow morning and get this in!

@ajratner I confirmed that the GPU test runs on my GPU machine and fails on my local mac due to no CUDA existing, so seems like it's doing what it should.  Give it a spin tomorrow.

File: metal/end_model/loss.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (bhancock8) [reply metal/end_model/loss.py]: I think there's a better way. torch.zeros() may default to cpu, but the torch.Tensor.new_zeros() will default to the same device as whatever tensor you call it on (same with .new_full()). So let's just have both cum_losses and cls_idx be on the same device as input using input.new_zeros() and input.new_full().

File: metal/multitask/mt_classifier.py [hunk 7]
```diff
@@ -42,19 +42,20 @@ def __init__(self, K, config):
...
         verbose=True,
+        print_confusion_matrix=False,
         **kwargs,
     ):
         """Scores the predictive performance of the Classifier on all tasks
         Args:
-            X: The input for the predict method
-            Y: A t-length list of [n] or [n, 1] np.ndarrays or torch.Tensors of
-                gold labels in {1,...,K_t}
+            data: either a Pytorch DataLoader or tuple supplying (X,Y):
```
Comment (bhancock8) [inline metal/multitask/mt_classifier.py]: Can't they also pass in a dataset here like they could before?

File: metal/utils.py [hunk 8]
```diff
@@ -365,3 +365,13 @@ def slice_data(data, indices):
             return outputs[0]
         else:
             return outputs
+
+
+def mt_to_cuda(data):
```
Comment (bhancock8) [inline metal/utils.py]: I think we need a better name here. I didn't now what "mt" meant. I mean, I assumed multitask, but what does it mean to convert a 'multitask' to cuda?

Can we also add a docstring here saying what the expected type of "data" is?

File: tests/gpu/test_gpu.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
+import os
+import pickle
+import unittest
+
+import GPUtil
```
Comment (bhancock8) [inline tests/gpu/test_gpu.py]: I'm glad we're not making this a general dependency for metal, since it (like our contrib stuff) is for a special use case and we don't want to have dependency bloat. Maybe as a help though, we could add a little tests/gpu/requirements.txt file (like we have in contrib/) so that people realize there are new requirements to run the code in this subdirectory.

File: metal/classifier.py [hunk 10]
```diff
@@ -270,45 +334,88 @@ def _set_scheduler(self, scheduler_config, optimizer):
...
+                update the class config.
+            print_confusion_matrix: Print confusion matrix
 
         Returns:
             scores: A (float) score or a list of such scores if kwarg metric
                 is a list
         """
-        Y = self._to_numpy(Y)
-        Y_p = self.predict(X, break_ties=break_ties, **kwargs)
+        Y_pred, Y = self._get_predictions(data, break_ties=break_ties, **kwargs)
```
Comment (bhancock8) [inline metal/classifier.py]: I understand the desire to have Y_pred instead of Y_p so that people don't confuse the p to mean "probabilistic", for example. But if we're going to make this change, I think we just need to be hard committed: all Y predictions everywhere must be Y_pred instead of Y_p. Looks like you changed it in a bunch of places, so that's probably what you're aiming for. If so, go ahead and update the design doc as well to reflect that. (And I'll add it as a TODO of mine to get the design doc out of a gdoc and into the repo so it's tracked). This change does get rid of the scheme for combining multiple single letter suffixes (e.g., Y_ps is predicted soft labels), but probably not a big deal, and I do like the clarity of Y_pred.

**Comment by bhancock8** [comment]
One more thought: how can the formatting issues be the fault of master @ajratner when it's passing the tests just fine? I think it's something in our additions. Is there a more verbose setting you can use to have it print out what exactly it's upset about or would be changing?

File: metal/classifier.py (hunk 10 above)
Comment (ajratner) [reply metal/classifier.py]: I'll revert to `Y_p` for now- can update terminology in a separate PR if we want.  Good catch!

Comment (ajratner) [reply metal/classifier.py]: Also another reason to have the design doc be tracked in the repo too- we should do that first!

File: metal/end_model/loss.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ajratner) [reply metal/end_model/loss.py]: Good call!

File: metal/utils.py (hunk 8 above)
Comment (ajratner) [reply metal/utils.py]: Also, this current function won't cover `X` that comes in as a list, right?  Will clean up now

**Comment by ajratner** [comment]
@bhancock8 I think it's because formatting is broken on master

**Comment by ajratner** [comment]
Passes GPU tests as well!
